File: examples/javascript/complex/middleware.js
const crypto = require('crypto');
const fs = require('fs');


function authMiddleware(req, res, next) {
    if (!req.session || !req.session.userId) {
        return res.status(401).json({ error: 'Authentication required' });
    }
    next();
}


function rateLimiter(windowMs, maxRequests) {
    const requests = new Map();

    return function(req, res, next) {
        const key = req.ip;
        const now = Date.now();
        const windowStart = now - windowMs;

        if (!requests.has(key)) {
            requests.set(key, []);
        }

        const timestamps = requests.get(key).filter(t => t > windowStart);
        timestamps.push(now);
        requests.set(key, timestamps);

        if (timestamps.length > maxRequests) {
            return res.status(429).json({ error: 'Too many requests' });
        }

        next();
    };
}


function corsMiddleware(req, res, next) {
    const origin = req.headers.origin;
    const allowedOrigins = (process.env.ALLOWED_ORIGINS || '*').split(',');
    if (allowedOrigins.includes('*') || allowedOrigins.includes(origin)) {
        res.setHeader('Access-Control-Allow-Origin', origin || '*');
    }
    res.setHeader('Access-Control-Allow-Methods', 'GET,POST,PUT,DELETE');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type,Authorization');
    if (req.method === 'OPTIONS') {
        return res.sendStatus(204);
    }
    next();
}


function errorHandler(err, req, res, next) {
    console.error('Error:', err.stack);
    const statusCode = err.statusCode || 500;
    const message = process.env.NODE_ENV === 'production'
        ? 'Internal Server Error'
        : err.message;
    res.status(statusCode).json({ error: message });
}


function requestLogger(req, res, next) {
    const start = Date.now();
    res.on('finish', () => {
        const duration = Date.now() - start;
        const logEntry = `${req.method} ${req.originalUrl} ${res.statusCode} ${duration}ms`;
        // VULN: Log injection - unsanitized user input in logs
        const userAgent = req.headers['user-agent'];
        console.log(`${logEntry} - ${userAgent}`);
    });
    next();
}


function csrfProtection(req, res, next) {
    if (['GET', 'HEAD', 'OPTIONS'].includes(req.method)) {
        return next();
    }
    const token = req.headers['x-csrf-token'] || req.body._csrf;
    const sessionToken = req.session.csrfToken;
    if (!token || token !== sessionToken) {
        return res.status(403).json({ error: 'CSRF token mismatch' });
    }
    next();
}


function generateCsrfToken() {
    return crypto.randomBytes(32).toString('hex');
}


function sanitizeInput(value) {
    if (typeof value !== 'string') return value;
    return value
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#x27;');
}


function validateContentType(expectedType) {
    return function(req, res, next) {
        const contentType = req.headers['content-type'];
        if (!contentType || !contentType.includes(expectedType)) {
            return res.status(415).json({ error: `Expected content type: ${expectedType}` });
        }
        next();
    };
}


module.exports = {
    authMiddleware,
    rateLimiter,
    corsMiddleware,
    errorHandler,
    requestLogger,
    csrfProtection,
    generateCsrfToken,
    sanitizeInput,
    validateContentType,
};

File: examples/go/complex/middleware.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"
)

type Middleware func(http.Handler) http.Handler

func AuthMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		token := r.Header.Get("Authorization")
		if token == "" {
			http.Error(w, "Unauthorized", 401)
			return
		}
		if !strings.HasPrefix(token, "Bearer ") {
			http.Error(w, "Invalid token format", 401)
			return
		}
		tokenValue := strings.TrimPrefix(token, "Bearer ")
		if !validateToken(tokenValue) {
			http.Error(w, "Invalid token", 401)
			return
		}
		next.ServeHTTP(w, r)
	})
}

func RateLimiter(maxRequests int, windowMs int64) Middleware {
	type clientInfo struct {
		timestamps []int64
		mu         sync.Mutex
	}
	clients := make(map[string]*clientInfo)
	var mu sync.Mutex

	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			clientIP := r.RemoteAddr
			mu.Lock()
			if _, exists := clients[clientIP]; !exists {
				clients[clientIP] = &clientInfo{}
			}
			client := clients[clientIP]
			mu.Unlock()

			client.mu.Lock()
			now := time.Now().UnixMilli()
			windowStart := now - windowMs
			var valid []int64
			for _, ts := range client.timestamps {
				if ts > windowStart {
					valid = append(valid, ts)
				}
			}
			valid = append(valid, now)
			client.timestamps = valid
			count := len(valid)
			client.mu.Unlock()

			if count > maxRequests {
				http.Error(w, "Too many requests", 429)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func CORSMiddleware(allowedOrigins []string) Middleware {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			origin := r.Header.Get("Origin")
			for _, allowed := range allowedOrigins {
				if allowed == "*" || allowed == origin {
					w.Header().Set("Access-Control-Allow-Origin", origin)
					break
				}
			}
			w.Header().Set("Access-Control-Allow-Methods", "GET,POST,PUT,DELETE")
			w.Header().Set("Access-Control-Allow-Headers", "Content-Type,Authorization")
			if r.Method == "OPTIONS" {
				w.WriteHeader(204)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func LoggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		duration := time.Since(start)
		// VULN: Log injection
		userAgent := r.Header.Get("User-Agent")
		fmt.Printf("REQUEST: %s %s %s %v\n", r.Method, r.URL.Path, userAgent, duration)
	})
}

func RecoveryMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if err := recover(); err != nil {
				fmt.Printf("PANIC: %v\n", err)
				http.Error(w, "Internal Server Error", 500)
			}
		}()
		next.ServeHTTP(w, r)
	})
}

func validateToken(token string) bool {
	secret := "gateway-secret-key"
	parts := strings.SplitN(token, ".", 2)
	if len(parts) != 2 {
		return false
	}
	payload := parts[0]
	signature := parts[1]
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(payload))
	expected := hex.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(signature), []byte(expected))
}

func generateToken(payload string) string {
	secret := "gateway-secret-key"
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(payload))
	signature := hex.EncodeToString(mac.Sum(nil))
	return payload + "." + signature
}

func main() {
	fmt.Println("Cloud API Gateway starting...")
}
